src/monster.cpp

```diff
@@ -1852,9 +1852,11 @@ void monster::add_msg_if_npc(const char *msg, ...) const
 {
     va_list ap;
     va_start(ap, msg);
-    std::string processed_npc_string = vstring_format(msg, ap);
-    processed_npc_string = replace_with_npc_name(processed_npc_string, disp_name());
-    add_msg(processed_npc_string.c_str());
+    if (g->u.sees(*this)) {
+        std::string processed_npc_string = vstring_format(msg, ap);
+        processed_npc_string = replace_with_npc_name(processed_npc_string, disp_name());
+        add_msg(processed_npc_string.c_str());
```

Comment:
Technically it should be `add_msg( "%s", ... )` because `add_msg` does string formating again, so format specifiers in the processed string will cause trouble, but that's a preexisting bug.

Practically, there should probably be a overload for `add_string` that takes a `std::string` and skips the formating.